### PR TITLE
Improve submodel ``idata`` creation

### PR DESCRIPTION
**Pull request by yannmclatchie**
## Description

This PR improves the method to build `InferenceData` objects from restricted models. We make the following changes in order to facilitate the creation of `idata` objects:

1. The PyMC `backend` object from the Bambi model is now stored in `ModelData` along with the ELPD of the model;
2. `_build_restricted_model` and `_build_data` methods are now class methods of the `Projector` class - the reasoning behind this is to restrict the need to pass large `ref_model` objects through many methods each time we project the reference model;
3. A `data/utils.py` module has been added for utility functions relating to data wrangling;
4. We compute the log-likelihood of projected models with the help of a `_compute_log_likelihood` method, currently found in the `data.utils.py`;
5. The dispersion parameter projections are transformed into their Bambi-defined transformations, e.g. in a linear regression we analytically project the `y_sigma` parameters and then transform them into `y_sigma_log__` in order to make use of this `_compute_log_likelihood` method.

Important considerations in the creation of a projected `idata` object are the preservation of dimensions. To this end, PyTorch tensors and sample dictionaries have been reshaped and manipulated into equivalent shapes. This has the knock-on effect of being able to compare the `ELPDData` objects and other diagnostic artefacts. 

### Open questions

It is unclear if there is a better way to deal with these Bambi-transformed parameters, since it is easy to define analytic projections of variance parameters and then transform them into into a reasonable space to be Bambi-compliant, but this is reliant on the end-user defining these dispersion parameters on standard spaces, e.g. `y_sigma` on the interval `[0,infty)`.

## Checklist

- [x] Tests covering the new functionality have been added
- [x] Documentation has been updated OR the change is too minor to be documented
- [x] Changes are listed in the `CHANGELOG.md` OR changes are insignificant

### CODE REVIEW

File: src/kulprit/base.py [hunk 1]
```diff
@@ -93,6 +100,98 @@ def __getitem__(self, model_size):  # pragma: no cover
...
+            X_res = torch.from_numpy(
+                np.column_stack(
+                    [
+                        self.ref_model.design.common[term]
+                        for term in restricted_common_terms
+                    ]
+                )
+            ).float()
+            # manually add intercept to new design matrix
+            X_res = torch.hstack((torch.ones(self.ref_model.num_obs, 1), X_res))
```
Comment (tomicapretto) [inline src/kulprit/base.py:139]: Before adding an intercept you should check whether the reference model actually contains one. You can do that with `bambi_model.intercept_term`. If it's not `None`, then the model contains an intercept. Otherwise, there's no intercept and you should do `restricted_common_terms = self.ref_model.common_terms[:(model_size + 1)]` or something equivalent.

File: src/kulprit/base.py [hunk 2]
```diff
@@ -93,6 +100,98 @@ def __getitem__(self, model_size):  # pragma: no cover
...
+        else:
+            # intercept-only model
+            X_res = torch.ones(self.ref_model.num_obs, 1).float()
+
+        # update common term names and dimensions and build new ModelData object
+        _, num_terms = X_res.shape
+        restricted_term_names = ["Intercept"] + restricted_common_terms
```
Comment (tomicapretto) [inline src/kulprit/base.py:146]: The same comment about the intercept applies here. First check if there's one.

File: src/kulprit/base.py [hunk 3]
```diff
@@ -93,6 +100,98 @@ def __getitem__(self, model_size):  # pragma: no cover
...
+        # build posterior dictionary from projected parameters
+        posterior = {
+            f"{term}": theta_perp[:, i] for i, term in enumerate(model.term_names)
+        }
+        if disp_perp is not None:
+            disp_dict = {
+                f"{model.response_name}_sigma": disp_perp,
+                f"{model.response_name}_sigma_log__": torch.log(disp_perp),
+            }
+            posterior.update(disp_dict)
```
Comment (tomicapretto) [inline src/kulprit/base.py:185]: Are you only considering models with a `sigma` parameter for now? I just wanted to point this would work for Gaussian linear regression, but it would fail with other cases. If you're ok with that, it's good!

File: src/kulprit/base.py [hunk 4]
```diff
@@ -27,6 +30,8 @@ def __init__(self, model, inferencedata=None):
         # build posterior if unavailable
         if inferencedata is None:
             inferencedata = model.fit()
+        # log the underlying backend model
+        backend = model.backend
```
Comment (tomicapretto) [inline src/kulprit/base.py:34]: I can't add the comment in its own line. But note in line 40 `term_names = list(model.term_names)` this gives the names of all the names (both common and group-specific).

File: src/kulprit/base.py (hunk 1 above)
Comment (tomicapretto) [reply src/kulprit/base.py:139]: Nevermind, I've just seen the class only works with models that have an intercept.

File: src/kulprit/base.py (hunk 2 above)
Comment (tomicapretto) [reply src/kulprit/base.py:146]: Nevermind.

File: src/kulprit/base.py [hunk 5]
```diff
@@ -54,14 +59,15 @@ def __init__(self, model, inferencedata=None):
         num_draws = (
             inferencedata.posterior.dims["chain"] * inferencedata.posterior.dims["draw"]
         )  # to do: test this for edge cases
-        # set the reference model's distance to itself as zero
+        # set the reference model's distance to itself as zero and compute ELPD
```
Comment (tomicapretto) [inline src/kulprit/base.py:62]: `num_obs, num_terms = model._design.common.design_matrix.shape` this may be problematic because the number of columns in the design matrix does not always match the number of terms in the model.

It's going to be true for regular numerical predictors. But if you have a categorical predictor, or something like a basis spline, it's not going to be true. This is because these types of terms span more than one column in the design matrix. I would use `len(model.terms)` or something similar.

Comment (tomicapretto) [reply src/kulprit/base.py:62]: On top of that, a cosmetic comment. I suggest using `idata` instead of `inferencedata`. It's going to save you a couple of lines of code because of line width rule and it's still readable because `idata` is a popular term.

Comment (tomicapretto) [reply src/kulprit/base.py:62]: What edge cases do you have in mind for 

```python
        num_draws = (
            inferencedata.posterior.dims["chain"] * inferencedata.posterior.dims["draw"]
        )  
```

I think it's good!

File: src/kulprit/base.py (hunk 1 above)
Comment (yannmclatchie) [reply src/kulprit/base.py:139]: Yeah in the future I hope to add compatibility with models without intercepts, but for now I will only deal with models that have an intercept 👍

File: src/kulprit/base.py (hunk 3 above)
Comment (yannmclatchie) [reply src/kulprit/base.py:185]: I have currently only implemented a Gaussian distribution for the variate, but later I will implement families without dispersion parameters (Binomial, Poisson, etc.)

File: src/kulprit/base.py (hunk 4 above)
Comment (yannmclatchie) [reply src/kulprit/base.py:34]: Yes this is how I have been using the `term_names` variable, but perhaps there is a better way we can consider in the future (or perhaps a better name we can give to the variable)

File: src/kulprit/base.py (hunk 5 above)
Comment (yannmclatchie) [reply src/kulprit/base.py:62]: `idata` is certainly cleaner, I'll make that change now! 

In terms of the more complex design matrix shapes, I haven't yet considered how to implement the procedure for any problem in which the latent predictor is the result from a regular linear regression. The categorical predictor is probably the most pressing issue, perhaps we can discuss later how to deal with it best?

File: src/kulprit/base.py [hunk 6]
```diff
@@ -93,6 +100,117 @@ def __getitem__(self, model_size):  # pragma: no cover
...
+
+        Returns:
+            arviz.inferencedata: Restricted model idata object
+        """
+
+        # reshape `theta_perp` so it has the same shape as the reference model
+        chain_n = len(self.ref_model.idata.posterior.coords.get("chain"))
+        draw_n = len(self.ref_model.idata.posterior.coords.get("draw"))
+        num_terms = model.num_terms
+        num_obs = self.ref_model.num_obs
```
Comment (tomicapretto) [inline src/kulprit/base.py:178]: Not strictly needed, but I think it's better if you have consistent names. I consider

* `chain_n`
* `draw_n`
* `terms_n`
* `obs_n`

to be more readable than what we have now. You could also use the `num_*` way if you like that more. But I recommend using it consistently (i.e. either all `*_n` or all `num_*`).

File: src/kulprit/base.py [hunk 7]
```diff
@@ -93,6 +100,117 @@ def __getitem__(self, model_size):  # pragma: no cover
...
+        idata_dims = (chain_n, draw_n, num_terms)
+        theta_perp = torch.reshape(theta_perp, idata_dims)
+
+        # build posterior dictionary from projected parameters
+        posterior = {
+            term: theta_perp[:, :, i] for i, term in enumerate(model.term_names)
+        }
+        if disp_perp is not None:
+            # reshape `disp_perp` if present
+            disp_perp_dims = (chain_n, draw_n, 1)
```
Comment (tomicapretto) [inline src/kulprit/base.py:188]: I'm not sure if the original dispersion parameter has shape of `(chain_n, draw_n, 1)`. I think it is just `(chain_n, draw_n)`. Could you check that?
However, if you want to concatenate this array with the array for the other parameters, well, it's right, you do need the third dimension.

File: tests/test_project.py [hunk 8]
```diff
@@ -62,16 +62,23 @@ def test_zero_model_size_project():
...
+def test_projected_idata_dims():
+    # extract dimensions of projected idata
+    res_model = proj.project(model_size=0)
+    idata_perp = res_model.idata
+    chain_n = len(idata_perp.posterior.coords.get("chain"))
+    draw_n = len(idata_perp.posterior.coords.get("draw"))
+    # ensure the restricted idata object has the same dimensions as that of the
+    # reference model
+    assert chain_n == len(idata.posterior.coords.get("chain"))
+    assert draw_n == len(idata.posterior.coords.get("draw"))
```
Comment (tomicapretto) [inline tests/test_project.py:84]: Maybe it's good to compare also the dimensions/coords for the number of observations and the dispersion parameter?

File: src/kulprit/base.py (hunk 6 above)
Comment (yannmclatchie) [reply src/kulprit/base.py:178]: Yes now that I see it I agree with you, although I do prefer the `num_*` format for verbosity so I'll change them to that as a matter of preference.

File: src/kulprit/base.py (hunk 7 above)
Comment (yannmclatchie) [reply src/kulprit/base.py:188]: There was no need for me to concatenate the tensors in the end, and so the extra dimension is indeed superfluous. I'll remove that now 👍

File: tests/test_project.py (hunk 8 above)
Comment (yannmclatchie) [reply tests/test_project.py:84]: I have added some tests for the dispersion parameter shape, although it turns out that I wasn't previously transferring the observed data to the projected model and thus couldn't test the number of observations in the projected `idata`. This is now being done in the `Projector._build_idata` method with 
```python
# add observed data component of projected idata
observed_data = {
    self.res_model.response_name = self.ref_model.idata.observed_data.get("y").to_dict().get("data")
}
```
and a test has been added.

**Comment by tomicapretto** [comment]
Congrats on this massive PR!
